driver/registry_memory.go

```diff
@@ -0,0 +1,308 @@
+package driver
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/gorilla/sessions"
+	"github.com/justinas/nosurf"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/ory/x/logrusx"
+
+	"github.com/ory/x/dbal"
+
+	"github.com/ory/hive-cloud/hive/selfservice"
+	"github.com/ory/hive-cloud/hive/selfservice/hooks"
+	"github.com/ory/hive-cloud/hive/selfservice/oidc"
+
+	"github.com/ory/herodot"
+
+	"github.com/ory/hive-cloud/hive/driver/configuration"
+	"github.com/ory/hive-cloud/hive/errorx"
+	"github.com/ory/hive-cloud/hive/identity"
+	password2 "github.com/ory/hive-cloud/hive/selfservice/password"
+	"github.com/ory/hive-cloud/hive/session"
+)
+
+var _ Registry = new(RegistryMemory)
+
+func init() {
+	dbal.RegisterDriver(new(RegistryMemory))
+}
+
+type RegistryMemory struct {
+	l             logrus.FieldLogger
+	ip            identity.Pool
+	ih            *identity.Handler
+	sm            session.Manager
+	sh            *session.Handler
+	em            errorx.Manager
+	eh            *errorx.Handler
+	c             configuration.Provider
+	ph            password2.Hasher
+	pv            password2.Validator
+	cookieManager sessions.Store
+
+	selfserviceRequestManager       selfservice.RequestManager
+	identityValidator               *identity.Validator
+	selfserviceRegistrationExecutor *selfservice.RegistrationExecutor
+	selfserviceLoginExecutor        *selfservice.LoginExecutor
+	selfserviceStrategyHandler      *selfservice.StrategyHandler
+	selfserviceStrategies           []selfservice.Strategy
+
+	writer herodot.Writer
+}
+
+func (m *RegistryMemory) WithLogger(l logrus.FieldLogger) Registry {
+	m.l = l
+	return m
+}
+
+func (m *RegistryMemory) NoSurf(router http.Handler) *nosurf.CSRFHandler {
+	n := nosurf.New(router)
+	n.SetFailureHandler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		m.Writer().WriteError(w, r, errors.WithStack(herodot.ErrBadRequest.WithReasonf("CSRF token is missing or invalid.")))
+	}))
+	return n
+}
+
+func (m *RegistryMemory) SelfServiceStrategies() []selfservice.Strategy {
+	if m.selfserviceStrategies == nil {
+		m.selfserviceStrategies = []selfservice.Strategy{
+			password2.NewStrategy(m, m.c),
+			oidc.NewStrategy(m, m.c),
+		}
+	}
+	return m.selfserviceStrategies
+}
+
+type postHooks []interface {
+	selfservice.HookLoginPostExecutor
+	selfservice.HookRegistrationPostExecutor
+}
+
+func (m *RegistryMemory) hooksPost(credentialsType identity.CredentialsType, configs []configuration.SelfServiceHook) postHooks {
+	var i postHooks
+
+	for _, hook := range configs {
+		switch hook.Run {
```

Comment:
Could these switch cases be further refactored into case-specific functions? A lot is happening in one switch.

Reply:
Yeah most likely